pytorch_lightning/loops/optimization/optimizer_loop.py

```diff
@@ -278,6 +279,8 @@ def _run_optimization(
         # ------------------------------
         # gradient update with accumulated gradients
         else:
+            if _fault_tolerant_training():
+                closure = _FaultToleranceClosureExecutor(closure, self.trainer)
```

Comment:
Can you explain this? Is it to avoid saving on exception if `optimizer.step` hasn't been called?

Why do we need it?
How is it related to the validation loop?

Reply:
This is to avoid saving a checkpoint on failure if the weights were being updated.